### ISSUE TITLE
Cannot assign to read only property 'Symbol(Symbol.hasInstance)' of function

### ISSUE THREAD

**Issue by rileyjshaw** [chrisguttandin/standardized-audio-context#986]
Hi! Thank you for all your work on this library.

I’m trying to use Tone.js on a project, which includes this library as a dependency. When I try to build my project using Babel I get the following error:

> TypeError: Cannot assign to read only property 'Symbol(Symbol.hasInstance)' of function 'function AudioBuffer(options){if(nativeOfflineAudioContextConstructor===null){throw new Error('Missing the nati...<omitted>...;}'

I initially logged the issue in [Tone.js’s repo](https://github.com/Tonejs/Tone.js/issues/731#issuecomment-1004337558), but I think the issue is actually here.

[Babel seems to prefer `defineProperty` when assigning `hasInstance`](https://github.com/babel/babel/issues/2286), and crashes otherwise. Would you accept a Pull Request changing the `Symbol.hasInstance` assignments in `audio-buffer-constructor.ts` and `periodic-wave-constructor.ts` to use `defineProperty` instead?

Thanks,

**Comment by rileyjshaw** [comment]
Since it’s such a small change, I just pushed an example PR to show what the change might look like: https://github.com/chrisguttandin/standardized-audio-context/pull/987

**Comment by chrisguttandin** [comment]
Hi @rileyjshaw, thanks for reporting this.

I'm a bit puzzled though since I'm using Babel as a direct or indirect build tool in each and every project and never ran into this error so far. Furthermore it looks like Babel's `createClass` helper is using `Obeject.defineProperty()` already. 

https://github.com/babel/babel/blob/ddd93b9c61a197711c2e3eae8a8b175783ea6976/packages/babel-helpers/src/helpers.ts#L219

Can you confirm that the fix in your PR solves the problem? Is there anything special in Gatsby's Babel configuration?

I noticed that it somehow configures the runtime but I don't fully understand what it actually does.

https://github.com/gatsbyjs/gatsby/blob/6ffe0afd15a4f4b327a108399f14183f337eab97/packages/babel-preset-gatsby/src/index.js#L118-L126

🤷‍♂️

**Comment by rileyjshaw** [comment]
> I noticed that it somehow configures the runtime but I don't fully understand what it actually does.

I’m with you - there are many layers of build tooling here 😅

It looks like Gatsby is pulling from the `build/es2019` directory, which preserves the `hasInstance` assignment. Rebuilding from my PR and replacing the `node_modules/standardized-audio-context/build` folder seems to fix the issue.

But… I’m encountering another build error from Tone.js, so I want to chase that down before I can fully endorse my PR. Initial exploration looks like my change helps, but hopefully I’ll be able to provide more detail soon.

I’m sorry that this issue report is so vague - this bug is turning into an unexpected deep dive into the Gatsby internals.

**Comment by rileyjshaw** [comment]
I’m heading out for tonight, but here are my notes so far:

- **The changes in PR #987 fix the error described in this issue.**
- To double-check, I imported the rebuilt `standardized-audio-context` directly into my project with no errors. The current release does throw an error.
- I’m encountering another issue from Tone.js which is unrelated, as far as I can tell.

---

This is unrelated to the issue at hand, but I’m leaving a bit more info on the Tone.js error for my own records below:

> TypeError: Cannot read properties of null (reading 'prototype')

it looks like Babel’s `_inheritsLoose` runtime helper is being called somewhere from Tone.js with a null argument. The function looks like this:

```
function _inheritsLoose(subClass, superClass) {
  subClass.prototype = Object.create(superClass.prototype);
  subClass.prototype.constructor = subClass;
  setPrototypeOf(subClass, superClass);
}
```

…so further down the rabbit hole we go 😂

**Comment by chrisguttandin** [comment]
Hi @rileyjshaw,

sorry for the long delay. I tried to replicate the bug today (mainly because I'm still scratching my head wondering how this works) but wasn't successful. I created a new Gatsby site following their tutorial over here: https://www.gatsbyjs.com/docs/quick-start/.

I then installed standardized-audio-context and added some nonsense code to the `index.js` file to make sure it gets loaded.

```js
import { AudioContext } from "standardized-audio-context";

if (typeof window !== 'undefined') {
    const audioContext = new AudioContext();

    alert(audioContext.state);
}
```

Gatsby's `build` and `develop` scripts both ran successfully. Is there anything else I need to do to trigger the bug?

**Comment by rileyjshaw** [comment]
Thanks for looking into it! I don’t know exactly what’s up with my repo, but I just tried a fresh install of it and encountered the same issue. I’ve mitigated the bug on my `main` branch, but just pushed a new branch for testing. If you have a spare moment, I think this will allow you to see what I’m seeing:

```
npm i -g gatsby-cli commitizen
git clone --single-branch --branch audio-context-bug https://github.com/rileyjshaw/rileyjshaw.github.io build-error-test
cd build-error-test
npm i
BABEL_ENV=test gatsby build # <-- I expect that it will fail here
# For subsequent builds…
npm run build
```

If standardized-audio-context builds without error on the default Gatsby starter, it must be an interaction with one of my project’s plugins 🤔

**Comment by rileyjshaw** [comment]
More clues: I encountered the initial bug while importing [a component using tone.js](https://github.com/rileyjshaw/rileyjshaw.github.io/blob/dev/src/components/scaling-faders-demo.js) into [a .mdx file](https://github.com/rileyjshaw/rileyjshaw.github.io/blob/dev/src/data/markdown/posts/published/2020-11-17-scaling-faders.mdx). The [gatsby-plugin-mdx](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-mdx) has its own build process, which appears to be the source of incompatibility. I imported the same component into a non-mdx page and built successfully, so I’m pretty confident it’s a weird interaction between MDX and this library.

At this point, this really seems like an edge case. I’ll leave it up to you weather you want to merge or close the issue / PR. Thank you for your time, and for maintaining this library!

**Comment by chrisguttandin** [comment]
I just read that a new version of MDX got released these days. Maybe that's going to fix the problem. https://mdxjs.com/blog/v2/

**Comment by rileyjshaw** [comment]
> I just read that a new version of MDX got released these days. Maybe that's going to fix the problem. https://mdxjs.com/blog/v2/

That’s right, and I hope it fixes the issue. Gatsby’s plugin isn’t updated to use V2 yet, but I’ll test it as soon as it is.

**Comment by chrisguttandin** [comment]
Do you have a (local and hacky) workaround that works for you right now or are you basically not able to build and deploy your project right now?

**Comment by rileyjshaw** [comment]
I do! Thanks for checking. I’m not blocked on this, and I might be the only person in the world experiencing it 🙃

**Comment by chrisguttandin** [comment]
In this case I would suggest that we keep this issue open for anyone else stumbling upon this. Hopefully it gets solved with MDX v2 in which case we can close it. And if it still remains even with v2 we need to somehow address it.

Does that sound good?

**Comment by rileyjshaw** [comment]
Sounds good to me! Thank you.

**Comment by chrisguttandin** [comment]
Hi @rileyjshaw, do you know if this is still an issue? It looks like the Gatsby plugin meanwhile got an update.

**Comment by rileyjshaw** [comment]
@chrisguttandin sorry for the delay; it’s taken me ages to get around to upgrading Gatsby to the latest version. This issue seems to be fixed now that Gatsby has moved to MDX v2  🎉  thanks for sticking with this, and for all the work you put into this library!
